README.md
[![PyPI version](https://badge.fury.io/py/styled.svg)](https://badge.fury.io/py/styled) ![PyPI - Python Version](https://img.shields.io/pypi/pyversions/styled.svg) [![Coverage Status](https://coveralls.io/repos/github/emdb-empiar/styled/badge.svg?branch=master)](https://coveralls.io/github/emdb-empiar/styled?branch=master) [![Build Status](https://travis-ci.org/emdb-empiar/styled.svg?branch=master)](https://travis-ci.org/emdb-empiar/styled)

# `styled`: Style your terminal with ease!

## Getting Started

To get it from `PyPI` use

```bash
pip install styled
```

It's best to do this in a virtual environment.

```bash
# anaconda/miniconda
conda create -n styled python
source activate styled
pip install styled

# virtualenv
virtualenv /path/to/env/styled -p /path/to/python
source /path/to/envs/styled/bin/activate
pip install styled

```

There are two ways to use `styled`:
* _interactively_ using the `styled` CLI which has three commands: `try`, `demo` and `version`; type `styled` to read the instructions;
* _programmatically_ through the `Styled` class. 

## Concepts
To style a string, you need to delimit the portion to be styled with double brackets `[[ ... ]]` then make sure that the following _three (3)_ conditions hold:

* separate the text from the styles with a pipe (`|`),
* quote the text part with either a pair of single (`'...'`) or double (`"..."`) quotes, then 
* separate each style with a colon (`:`)

There are _three (3)_ types of styles:

* _text styling_ such as `bold`, `blink`, `underlined` etc.
* *foreground colours*, such as `fg-red`,
* *background colours*, such as `bg-blue`.

If you want to style an extended piece of text you can use the special style marker `no-end` signaling that current style will not be interrupted for as long as possible. This can be useful if you want to style a long stretch of text e.g. a paragraph. Use the 
special marker `yes-end` to terminate a `no-end` marker.

## Interactive Use
The fastest way to get started is to use the `try` command:

![Trying your hand out of a string of formatted text](try.png) _Trying out a formatted string._

You can reference the available styles using the `demo` command, which by default shows colours.

![Viewing the colour palette](colours.png) _Colour Palette_

Use `demo formats` to view formatting options.

![Viewing the formatting options](formats.png) _Formatting Options_

## Programmatic Use

To use `styled` programmatically, use the `Styled` class as follows:

```python
from styled import Styled
s = Styled("We are [[ 'bold'|bold ]] men!")
print(s)
```

You can perform string formatting _directly_ in the constructor.

```python
from styled import Styled
s = Styled("There were up to [[ '{}'|bold:fg-red ]] people who handed over themselves to the \
[[ '{police}'|fg-black:bg-red:bold ]].", 24, police='policia')
print(s)
```

```python
from styled import Styled
s = Styled("There were up to [[ '{}'|bold:fg-red:no-end ]] people who handed over themselves to the \
[[ '{police}'|fg-black:bg-red:bold ]].", 24, police='policia')
print(s)
``` 

The above example will have all the text until the next marker affected by the red foreground styling. You can also 
manually enforce an end marker by using `yes-end` as a style. By default, all style markers will terminate on the 
text to be styled. So, for example

```python
from styled import Styled
# an example of a terminating end marker
s = Styled("There were up to [[ '{}'|bold:fg-red:no-end ]] people who handed over themselves [[ ''|yes-end ]] to the \
[[ '{police}'|fg-black:bg-red:bold ]].", 24, police='policia')
print(s)
``` 

will suspend red foreground at the end of the word 'themselves'.

You can only have one foreground and one background colour. Ignoring this produces a `StyleError`

```python
from styled import Styled
s = Styled("There were up to [[ '{}'|bold:fg-red:fg-blue ]] people who handed over themselves to the [[ '{police}'|fg-black:bg-red:bold ]].", 24, police='policia')
```

```shell
Traceback (most recent call last):
  File "/Users/pkorir/miniconda2/envs/styled/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2878, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-0993b680f88b>", line 1, in <module>
    s = Styled("There were up to [[ '{}'|bold:fg-red:fg-blue ]] people who handed over themselves to the [[ '{police}'|fg-black:bg-red:bold ]].", 24, police='policia')
  File "/Users/pkorir/PycharmProjects/styled/styled/styled.py", line 55, in __init__
    self._validate(self._tokens)
  File "/Users/pkorir/PycharmProjects/styled/styled/styled.py", line 156, in _validate
    raise StyleError(u"Multiple foreground styles for text '{}': {}".format(text, ', '.join(styles)))
StyleError: Multiple foreground styles for text '24': bold, fg-red, fg-blue
```

Inputting an invalid `style` also raises a `StyleError`

```python
from styled import Styled
s = Styled("This just can't just [[ 'go'|underline ]] on forever! ")
```

```shell
Traceback (most recent call last):
  File "/Users/pkorir/miniconda2/envs/styled/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2878, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-528d6d2ac4f4>", line 1, in <module>
    s = Styled("This just can't just [[ 'go'|underline ]] on forever! ")
  File "/Users/pkorir/PycharmProjects/styled/styled/styled.py", line 59, in __init__
    self._styled = self._transform(self._plain, self._cleaned_tokens)
  File "/Users/pkorir/PycharmProjects/styled/styled/styled.py", line 99, in _transform
    styled += plain[i:start] + self.transform(token)
  File "/Users/pkorir/PycharmProjects/styled/styled/styled.py", line 87, in transform
    raise StyleError(u"Unknown style '{}'".format(style_))
StyleError: Unknown style 'underline'
```

(In case you're wondering, it should have been `underlined` not `underline`.)


## Concatenation

Concatenating a normal string and a `Styled` string produces a `Styled` string, which is a subclass of string. 
Internally, though, everything is a Unicode string.

```shell
from styled import Styled
s = Styled("This just can't just [[ 'go'|underlined ]] on forever! ")
u = "She shouted back!"
print(type(s + u))
# <class 'styled.styled.Styled'>
print(type(u + s))
# <class 'styled.styled.Styled'>
s += "Gloria!"
print(type(s))
# <class 'styled.styled.Styled'>
```

## Validation

The process of generating the output involves some validation - to check that styles are sane. At present,
only multiple fore- and background colours are checked. This will be expanded as needed.

## Cleaning Styles

In addition to validation, styles are cleaned. Cleaning ensures that the final set of styles is non-redundant.

```python
from styled import Styled
s = Styled("It takes enormous [[ 'courage'|bold:bold:bold ]] to admit that you're wrong.")
print(s._tokens)
# [(19, 49, u'courage', [u'bold', u'bold', u'bold'])]
print(s._cleaned_tokens)
# [(19, 49, u'courage', [u'bold'])]
```

## Backstory
Welcome to `styled`, a simple Python package that makes a breeze of writing beautiful text to the terminal. 
The main innovation behind `styled` is the dead-simple user interface which does away with the user's need to know 
anything other than the style names. Behind the scenes `styled` handles everything to keep 
your styles consistent and redundant and informing you when you have made formatting errors.

`styled` was borne out of the frustration encountered in using other packages which muddle the boundary between
_user-space_ and _design-space_. The user should be free to be a _user_, and it is the _designer's_ job to hide the 
implementation behind a simple user interface that facilitates the user's task. This is what I've tried to do. If I have 
failed to live up to this please let me know. I'm sure together we can come up with something better.

## Aspirations?

When I grow up I want to have my own Python string declaration like so:

```shell
# hey! I'm a styled string
s = s"You have to [[ 'believe'|fg-red ]] it to [[ 'see'|fg-green ]] it!"
```

## Special Thanks

To the following people

* **Dimitris Zlatanidis** (for the inspiration and resources in his package `colored` available at <https://gitlab.com/dslackw/colored>)
* **Cesare Catavitello** (for being the _first_ user)

CHANGELOG.md
# History

## 0.3.1 [2023-10-26]

* Bugfix: tentative fix to handle string with braces e.g. `yl-4-({4-[(4-methylbenzene-1-sulfonyl)oxy]phenyl}diazenyl)-1H-pyrazol-1-yl`
* Dropped support for Python 3.5 and 3.6
* Added support for Python 3.12
* Added this changelog


styled/tests.py
# -*- coding: utf-8 -*-
from __future__ import print_function
from unittest import TestCase

from .styled import Styled, StyleError


class NonStringType(object):
    def __init__(self, s):
        self.s = s


class TestStyled(TestCase):
    def test_default(self):
        string = 'this is a string'
        S = Styled(string)
        self.assertEqual(S, string)
        self.assertIsInstance(S, Styled)

    def test_empty(self):
        s = Styled()
        self.assertEqual(len(s), 0)

    def test_type(self):
        s = NonStringType(27)
        with self.assertRaises(ValueError):
            Styled(s)

    def test_find_tokens(self):
        s = """[[ 'a word'|fg-red ]]"""
        s = Styled(s)
        self.assertItemsEqual(s._tokens, [(0, 21, u'a word', [u'fg-red'])])
        s = Styled("""[[ 'your {} is open'|fg-blue ]]""", 'bank')
        self.assertItemsEqual(s._tokens, [(0, 33, u'your bank is open', [u'fg-blue'])])

    def test_length(self):
        u = 'I am the most handsome guy in the room.'
        u_list = u.split(' ')
        s = Styled(" ".join(u_list[:4] + ["[[ '{}'|bold ]]".format(u_list[4])] + u_list[5:]))
        self.assertEqual(len(u), len(s))

    def test_format(self):
        u = "This is a very {} affair in which {count} people were involved."
        s = Styled(u, 'noble', count=38)
        self.assertEqual(s, u.format('noble', count=38))

    def test_quotes(self):
        sq = Styled("I have a [[ 'bold \"hair\"'|fg-red ]] face.")
        dq = Styled('I have a [[ "bold \'hair\'"|fg-red ]] face.')
        tq = Styled("""I have a [[ "bold 'hair'"|fg-red ]] face.""")
        self.assertItemsEqual(sq._tokens, [(9, 35, u'bold "hair"', [u'fg-red'])])
        self.assertItemsEqual(dq._tokens, [(9, 35, u"bold 'hair'", [u'fg-red'])])
        self.assertItemsEqual(tq._tokens, [(9, 35, u"bold 'hair'", [u'fg-red'])])

    def test_style_error(self):
        with self.assertRaises(StyleError):
            Styled('I have a [[ "bold\'hair|fg-red ]] face')

    def test_concatenate(self):
        s1 = Styled("This is the [[ 'end'|bold ]]!")
        u1 = " And I will be finished."
        c1 = s1 + u1
        self.assertEqual(c1, s1 + u1)
        self.assertIsInstance(c1, Styled)
        u2 = "I will be finished..."
        s2 = Styled(" when this [[ 'ends'|bold ]]!")
        c2 = u2 + s2
        self.assertEqual(c2, u2 + s2)
        self.assertIsInstance(c2, Styled)

    def test_fg_colour(self):
        s = Styled("I have never seen a [[ 'white'|fg-white:bold ]] stallion.")
        self.assertIsInstance(s, Styled)

    def test_bg_colour(self):
        s = Styled("[[ 'White bold text on a black background.'|fg-black:bg-white:bold ]]")
        self.assertIsInstance(s, Styled)

    def test_catch_multiple_fgs(self):
        with self.assertRaises(StyleError):
            s = Styled("[[ 'useless'|fg-red:fg-orange ]]")

    def test_catch_multiple_bgs(self):
        with self.assertRaises(StyleError):
            s = Styled("[[ 'useless'|bg-red:bg-orange ]]")

    def test_clean_tokens(self):
        s = Styled("[[ 'some text'|bold:bold ]]")
        self.assertItemsEqual(s._cleaned_tokens, [(0, 27, u"some text", [u'bold'])])

    def test_iteration(self):
        s = Styled("There are some folks who [[ 'gasp'|fg-black:bg-deep_sky_blue_2:underlined:blink ]] at the thought "
                   "of the military. ")
        s += 'Woe unto them!'
        self.assertIsInstance(s, Styled)

    def test_unicode(self):
        s = Styled("We wish we had [[ 'red'|fg-red ]] faces")
        u_s = unicode(s)
        s_s = str(s)
        u_ = u"Thërę are some folkß who [[ 'gæsp'|fg-black:bg-deep_sky_blue_2:underlined:blink ]] at the " \
             u"thœught of the military. "
        e = Styled(u_)
        print(e)
        self.assertIsInstance(u_s, unicode)
        self.assertIsInstance(s_s, str)
        self.assertIsInstance(s, Styled)
        self.assertIsInstance(e, Styled)

    def test_concat_unicode(self):
        u_s = u"A unicode string"
        s = Styled("A [[ 'styled'|blink ]] string")
        c1 = u_s + s
        c2 = s + u_s
        self.assertIsInstance(u_s, unicode)
        self.assertIsInstance(c1, Styled)
        self.assertIsInstance(c2, Styled)




tests/test_styled.py
# -*- coding: utf-8 -*-
from __future__ import print_function

import sys
from unittest import TestCase

from styled import Styled, StyleError, ESC, END


class NonStringType(object):
    def __init__(self, s):
        self.s = s


class Py23Fix(object):
    def __init__(self, *args, **kwargs):
        if sys.version_info[0] > 2:
            pass
        else:
            # new names for assert methods
            self.assertCountEqual = self.assertItemsEqual
        super(Py23Fix, self).__init__(*args, **kwargs)


class BaseTestStyled(Py23Fix, TestCase):
    pass


class TestStyled(BaseTestStyled):
    def test_default(self):
        string = 'this is a string'
        S = Styled(string)
        self.assertEqual(S, string)
        self.assertIsInstance(S, Styled)

    def test_empty(self):
        s = Styled()
        self.assertEqual(len(s), 0)

    def test_type(self):
        s = NonStringType(27)
        with self.assertRaises(ValueError):
            Styled(s)

    def test_find_tokens(self):
        s = """[[ 'a word'|fg-red ]]"""
        s = Styled(s)
        self.assertCountEqual(s._tokens, [(0, 21, u'a word', [u'fg-red'])])
        s = Styled("""[[ 'your {} is open'|fg-blue ]]""", 'bank')
        self.assertCountEqual(s._tokens, [(0, 33, u'your bank is open', [u'fg-blue'])])

    def test_length(self):
        u = 'I am the most handsome guy in the room.'
        u_list = u.split(' ')
        s = Styled(" ".join(u_list[:4] + ["[[ '{}'|bold ]]".format(u_list[4])] + u_list[5:]))
        self.assertEqual(len(u), len(s))

    def test_format(self):
        u = "This is a very {} affair in which {count} people were involved."
        s = Styled(u, 'noble', count=38)
        self.assertEqual(s, u.format('noble', count=38))

    def test_quotes(self):
        sq = Styled("I have a [[ 'bold \"hair\"'|fg-red ]] face.")
        dq = Styled('I have a [[ "bold \'hair\'"|fg-red ]] face.')
        tq = Styled("""I have a [[ "bold 'hair'"|fg-red ]] face.""")
        self.assertCountEqual(sq._tokens, [(9, 35, u'bold "hair"', [u'fg-red'])])
        self.assertCountEqual(dq._tokens, [(9, 35, u"bold 'hair'", [u'fg-red'])])
        self.assertCountEqual(tq._tokens, [(9, 35, u"bold 'hair'", [u'fg-red'])])

    def test_style_error(self):
        with self.assertRaises(StyleError):
            Styled('I have a [[ "bold\'hair|fg-red ]] face')

    def test_invalid_style(self):
        with self.assertRaises(StyleError):
            Styled("I have a [[ 'good'|fg-sldjdlj ]]")

    def test_contatenate_styled(self):
        s1 = Styled("This is the [[ 'end'|bold ]]!")
        s2 = Styled(" when this [[ 'ends'|bold ]]!")
        s3 = s1 + s2
        self.assertIsInstance(s3, Styled)
        self.assertEqual(s3, s1 + s2)
        self.assertEqual(len(s3), len(s1) + len(s2))

    def test_concatenate_with_other(self):
        s1 = Styled("This is the [[ 'end'|bold ]]!")
        u1 = " And I will be finished."
        c1 = s1 + u1
        self.assertEqual(c1, s1 + u1)
        self.assertIsInstance(c1, Styled)
        u2 = "I will be finished..."
        s2 = Styled(" when this [[ 'ends'|bold ]]!")
        c2 = u2 + s2
        self.assertEqual(c2, u2 + s2)
        self.assertIsInstance(c2, Styled)

    def test_fg_colour(self):
        s = Styled("I have never seen a [[ 'white'|fg-white:bold ]] stallion.")
        self.assertIsInstance(s, Styled)

    def test_bg_colour(self):
        s = Styled("[[ 'White bold text on a black background.'|fg-black:bg-white:bold ]]")
        self.assertIsInstance(s, Styled)

    def test_catch_multiple_fgs(self):
        with self.assertRaises(StyleError):
            s = Styled("[[ 'useless'|fg-red:fg-orange ]]")

    def test_catch_multiple_bgs(self):
        with self.assertRaises(StyleError):
            s = Styled("[[ 'useless'|bg-red:bg-orange ]]")

    def test_catch_multiple_no_ends(self):
        with self.assertRaises(StyleError):
            s = Styled("[[ 'useless'|bg-red:no-end:no-end ]]")

    def test_clean_tokens(self):
        s = Styled("[[ 'some text'|bold:bold ]]")
        # self.assertItemsEqual(s._cleaned_tokens, [(0, 27, u"some text", [u'bold'])])
        self.assertCountEqual(s._cleaned_tokens, [(0, 27, u"some text", [u'bold'])])

    def test_iteration(self):
        s = Styled("There are some folks who [[ 'gasp'|fg-black:bg-deep_sky_blue_2:underlined:blink ]] at the thought "
                   "of the military. ")
        s += 'Woe unto them!'
        self.assertIsInstance(s, Styled)

    def test_unicode(self):
        s = Styled("We wish we had [[ 'red'|fg-red ]] faces")
        if sys.version_info[0] > 2:
            u_s = str(s)
        else:
            u_s = unicode(s)
        s_s = str(s)
        u_ = u"Thërę are some folkß who [[ 'gæsp'|fg-black:bg-deep_sky_blue_2:underlined:blink ]] at the " \
             u"thœught of the military. "
        e = Styled(u_)
        if sys.version_info[0] > 2:
            self.assertIsInstance(u_s, str)
        else:
            self.assertIsInstance(u_s, unicode)
        self.assertIsInstance(s_s, str)
        self.assertIsInstance(s, Styled)
        self.assertIsInstance(e, Styled)

    def test_concat_unicode(self):
        u_s = u"A unicode string"
        s = Styled("A [[ 'styled'|blink ]] string")
        c1 = u_s + s
        c2 = s + u_s
        if sys.version_info[0] > 2:
            self.assertIsInstance(u_s, str)
        else:
            self.assertIsInstance(u_s, unicode)
        self.assertIsInstance(c1, Styled)
        self.assertIsInstance(c2, Styled)

    def test_no_end(self):
        # lacks end
        u_ = u"[[ 'gæsp'|fg-black:bg-deep_sky_blue_2:underlined:blink:no-end ]]"
        e = Styled(u_)
        self.assertNotEqual(e[-3:], u'[0m'.format(ESC, END))
        # has end
        u_ = u"[[ 'gæsp'|fg-black:bg-deep_sky_blue_2:underlined:blink ]]"
        e = Styled(u_)
        self.assertEqual(e[-3:], u'[0m'.format(ESC, END))

    def test_yes_end(self):
        u_ = u"[[ 'wonder-woman'|fg-red:no-end ]] and this text will also be red [[ 'super-man'|fg-blue:yes-end ]]" \
             u"and this text will be normal"
        e = Styled(u_)
        # test that there is not styling at the end
        self.assertTrue(e[-1], u_[-1])


styled/core.py
# -*- coding: utf-8 -*-
from __future__ import print_function

import re

ESC = u'\033'
# ESC = u'\xb1'
# ESC = u'\e'

STYLES = {
    u'bold': u"1",
    u'underlined': u"4",
    u'red': u'38;5;1',
    u'blink': u"5",
    u'reverse': u"7",
    u"green": u"38;5;2",
    u"yellow": u"38;5;3",
    u"blue": u"38;5;4",
    u"magenta": u"38;5;5",
    u"cyan": u"38;5;6",
    u"light_gray": u"38;5;7",
    u"dark_gray": u"38;5;8",
    u"light_red": u"38;5;9",
    u"light_green": u"38;5;10",
    u"light_yellow": u"38;5;11",
    u"light_blue": u"38;5;12",
    u"light_magenta": u"38;5;13",
    u"light_cyan": u"38;5;14",

}

END = u'{}[0m'.format(ESC)


class Styled(str):
    pattern = re.compile(r".*?(?P<pattern>[[][[].*?[|].*?[]][]]).*?")
    styled_text = re.compile(r".*?[[][[].*?'(?P<text>.*?)'[|](?P<styles>(\w+[:]?)+).*?[]][]].*")

    # todo: make plain and styled readonly attributes that are only set at construction time
    def __new__(cls, s, *args, **kwargs):
        obj = super(Styled, cls).__new__(cls, s)
        # format string using args and kwargs
        cls._plain = obj.format(*args, **kwargs)
        # extract tokens
        cls._tokens = obj._find_tokens(cls._plain)
        # transform text with tokens and styles
        cls._styled = obj._transform(cls._plain, cls._tokens)
        cls._unstyled = obj._transform(cls._plain, cls._tokens, apply=False)
        # cls.finds = cls.parse_styles(cls._plain)
        return obj

    def __init__(self, s, *args, **kwargs):
        super(Styled, self).__init__(s)

    @staticmethod
    def transform(token):
        """Static method that converts tokens into styled text"""
        start, end, text, styles = token
        s = u''
        for style in styles:
            s += u'{}[{}m'.format(ESC, STYLES[style])
        return u'{}{}{}'.format(s, text, END)

    @classmethod
    def _transform(cls, plain, tokens, apply=True):
        """Transform the whole string into a styled string"""
        i = 0
        sss = u''
        for token in tokens:
            start, end, text, styles = token
            if apply:
                sss += plain[i:start] + cls.transform(token)
            else:
                sss += plain[i:start] + text
            i = end
        sss += plain[i:]
        return sss

    @classmethod
    def _find_tokens(cls, string):
        """Find all style tokens in the string"""
        tokens = list()
        index = 0
        pos = 0
        while True:
            string = string[index:]
            result = cls.pattern.match(string)
            if not result:  # or not result2:
                break
            find = result.group(u'pattern')
            result2 = cls.styled_text.match(find)
            text = result2.group(u'text')
            styles = result2.group(u'styles').split(u':')
            t_ = (result.start() + pos + (result.end() - len(find)), result.end() + pos, text, styles)
            tokens.append(
                t_,
            )
            index = result.end()
            pos += index
        return tokens

    def __len__(self):
        return len(self._unstyled)

    # @classmethod
    # def parse_styles(cls, string):
    #     finds = list()
    #     index = 0
    #     pos = 0
    #     while True:
    #         # print('string = ', string)
    #         string = string[index:]
    #         result = cls.pattern.match(string)
    #         # result2 = cls.styled_text.match(string)
    #         if not result:  # or not result2:
    #             break
    #         find = result.group('pattern')
    #         result2 = cls.styled_text.match(find)
    #         text = result2.group('text')
    #         # text = result.group('text')
    #         styles = result2.group('styles').split(':')
    #         # styles = result.group('styles').split(':')
    #         f_ = (result.start() + pos + (result.end() - len(find)), result.end() + pos, text, styles)
    #         # print(f_)
    #         finds.append(
    #             f_,
    #         )
    #         index = result.end()
    #         pos += index
    #         # print()
    #
    #     def transform(find):
    #         _, __, text, styles = find
    #         s = ''
    #         for style in styles:
    #             s += '{}[{}m'.format(ESC, STYLES[style])
    #         return '{}{}{}'.format(s, text, END)
    #
    #     i = 0
    #     sss = ''
    #     for find in finds:
    #         a, b, c, d = find
    #         sss += cls._plain[i:a] + transform(find)
    #         print(find, transform(find))
    #         i = b
    #     sss += cls._plain[i:]
    #     print(sss)
    #
    #     return finds


styled/assets.py
# -*- coding: utf-8 -*-
"""Module defining colour and style assets"""

ESC = u'\x1b'

END = u'{}[0m'.format(ESC)

STYLE_NAMES = {
    u'bold': u"1",
    u'dim': u"2",
    u'underlined': u"4",
    u'blink': u"5",
    u'reverse': u"7",
    u'hidden': u"8",
    u'reset': u"0",
    u'res_bold': u"21",
    u'res_dim': u"22",
    u'res_underlined': u"24",
    u'res_blink': u"25",
    u'res_reverse': u"27",
    u'res_hidden': u"28"
}

COLOURS = {
    u"black": u"0",
    u"red": u"1",
    u"green": u"2",
    u"yellow": u"3",
    u"blue": u"4",
    u"magenta": u"5",
    u"cyan": u"6",
    u"light_gray": u"7",
    u"dark_gray": u"8",
    u"light_red": u"9",
    u"light_green": u"10",
    u"light_yellow": u"11",
    u"light_blue": u"12",
    u"light_magenta": u"13",
    u"light_cyan": u"14",
    u"white": u"15",
    u"grey_0": u"16",
    u"navy_blue": u"17",
    u"dark_blue": u"18",
    u"blue_3a": u"19",
    u"blue_3b": u"20",
    u"blue_1": u"21",
    u"dark_green": u"22",
    u"deep_sky_blue_4a": u"23",
    u"deep_sky_blue_4b": u"24",
    u"deep_sky_blue_4c": u"25",
    u"dodger_blue_3": u"26",
    u"dodger_blue_2": u"27",
    u"green_4": u"28",
    u"spring_green_4": u"29",
    u"turquoise_4": u"30",
    u"deep_sky_blue_3a": u"31",
    u"deep_sky_blue_3b": u"32",
    u"dodger_blue_1": u"33",
    u"green_3a": u"34",
    u"spring_green_3a": u"35",
    u"dark_cyan": u"36",
    u"light_sea_green": u"37",
    u"deep_sky_blue_2": u"38",
    u"deep_sky_blue_1": u"39",
    u"green_3b": u"40",
    u"spring_green_3b": u"41",
    u"spring_green_2a": u"42",
    u"cyan_3": u"43",
    u"dark_turquoise": u"44",
    u"turquoise_2": u"45",
    u"green_1": u"46",
    u"spring_green_2b": u"47",
    u"spring_green_1": u"48",
    u"medium_spring_green": u"49",
    u"cyan_2": u"50",
    u"cyan_1": u"51",
    u"dark_red_1": u"52",
    u"deep_pink_4a": u"53",
    u"purple_4a": u"54",
    u"purple_4b": u"55",
    u"purple_3": u"56",
    u"blue_violet": u"57",
    u"orange_4a": u"58",
    u"grey_37": u"59",
    u"medium_purple_4": u"60",
    u"slate_blue_3a": u"61",
    u"slate_blue_3b": u"62",
    u"royal_blue_1": u"63",
    u"chartreuse_4": u"64",
    u"dark_sea_green_4a": u"65",
    u"pale_turquoise_4": u"66",
    u"steel_blue": u"67",
    u"steel_blue_3": u"68",
    u"cornflower_blue": u"69",
    u"chartreuse_3a": u"70",
    u"dark_sea_green_4b": u"71",
    u"cadet_blue_2": u"72",
    u"cadet_blue_1": u"73",
    u"sky_blue_3": u"74",
    u"steel_blue_1a": u"75",
    u"chartreuse_3b": u"76",
    u"pale_green_3a": u"77",
    u"sea_green_3": u"78",
    u"aquamarine_3": u"79",
    u"medium_turquoise": u"80",
    u"steel_blue_1b": u"81",
    u"chartreuse_2a": u"82",
    u"sea_green_2": u"83",
    u"sea_green_1a": u"84",
    u"sea_green_1b": u"85",
    u"aquamarine_1a": u"86",
    u"dark_slate_gray_2": u"87",
    u"dark_red_2": u"88",
    u"deep_pink_4b": u"89",
    u"dark_magenta_1": u"90",
    u"dark_magenta_2": u"91",
    u"dark_violet_1a": u"92",
    u"purple_1a": u"93",
    u"orange_4b": u"94",
    u"light_pink_4": u"95",
    u"plum_4": u"96",
    u"medium_purple_3a": u"97",
    u"medium_purple_3b": u"98",
    u"slate_blue_1": u"99",
    u"yellow_4a": u"100",
    u"wheat_4": u"101",
    u"grey_53": u"102",
    u"light_slate_grey": u"103",
    u"medium_purple": u"104",
    u"light_slate_blue": u"105",
    u"yellow_4b": u"106",
    u"dark_olive_green_3a": u"107",
    u"dark_green_sea": u"108",
    u"light_sky_blue_3a": u"109",
    u"light_sky_blue_3b": u"110",
    u"sky_blue_2": u"111",
    u"chartreuse_2b": u"112",
    u"dark_olive_green_3b": u"113",
    u"pale_green_3b": u"114",
    u"dark_sea_green_3a": u"115",
    u"dark_slate_gray_3": u"116",
    u"sky_blue_1": u"117",
    u"chartreuse_1": u"118",
    u"light_green_2": u"119",
    u"light_green_3": u"120",
    u"pale_green_1a": u"121",
    u"aquamarine_1b": u"122",
    u"dark_slate_gray_1": u"123",
    u"red_3a": u"124",
    u"deep_pink_4c": u"125",
    u"medium_violet_red": u"126",
    u"magenta_3a": u"127",
    u"dark_violet_1b": u"128",
    u"purple_1b": u"129",
    u"dark_orange_3a": u"130",
    u"indian_red_1a": u"131",
    u"hot_pink_3a": u"132",
    u"medium_orchid_3": u"133",
    u"medium_orchid": u"134",
    u"medium_purple_2a": u"135",
    u"dark_goldenrod": u"136",
    u"light_salmon_3a": u"137",
    u"rosy_brown": u"138",
    u"grey_63": u"139",
    u"medium_purple_2b": u"140",
    u"medium_purple_1": u"141",
    u"gold_3a": u"142",
    u"dark_khaki": u"143",
    u"navajo_white_3": u"144",
    u"grey_69": u"145",
    u"light_steel_blue_3": u"146",
    u"light_steel_blue": u"147",
    u"yellow_3a": u"148",
    u"dark_olive_green_3": u"149",
    u"dark_sea_green_3b": u"150",
    u"dark_sea_green_2": u"151",
    u"light_cyan_3": u"152",
    u"light_sky_blue_1": u"153",
    u"green_yellow": u"154",
    u"dark_olive_green_2": u"155",
    u"pale_green_1b": u"156",
    u"dark_sea_green_5b": u"157",
    u"dark_sea_green_5a": u"158",
    u"pale_turquoise_1": u"159",
    u"red_3b": u"160",
    u"deep_pink_3a": u"161",
    u"deep_pink_3b": u"162",
    u"magenta_3b": u"163",
    u"magenta_3c": u"164",
    u"magenta_2a": u"165",
    u"dark_orange_3b": u"166",
    u"indian_red_1b": u"167",
    u"hot_pink_3b": u"168",
    u"hot_pink_2": u"169",
    u"orchid": u"170",
    u"medium_orchid_1a": u"171",
    u"orange_3": u"172",
    u"light_salmon_3b": u"173",
    u"light_pink_3": u"174",
    u"pink_3": u"175",
    u"plum_3": u"176",
    u"violet": u"177",
    u"gold_3b": u"178",
    u"light_goldenrod_3": u"179",
    u"tan": u"180",
    u"misty_rose_3": u"181",
    u"thistle_3": u"182",
    u"plum_2": u"183",
    u"yellow_3b": u"184",
    u"khaki_3": u"185",
    u"light_goldenrod_2a": u"186",
    u"light_yellow_3": u"187",
    u"grey_84": u"188",
    u"light_steel_blue_1": u"189",
    u"yellow_2": u"190",
    u"dark_olive_green_1a": u"191",
    u"dark_olive_green_1b": u"192",
    u"dark_sea_green_1": u"193",
    u"honeydew_2": u"194",
    u"light_cyan_1": u"195",
    u"red_1": u"196",
    u"deep_pink_2": u"197",
    u"deep_pink_1a": u"198",
    u"deep_pink_1b": u"199",
    u"magenta_2b": u"200",
    u"magenta_1": u"201",
    u"orange_red_1": u"202",
    u"indian_red_1c": u"203",
    u"indian_red_1d": u"204",
    u"hot_pink_1a": u"205",
    u"hot_pink_1b": u"206",
    u"medium_orchid_1b": u"207",
    u"dark_orange": u"208",
    u"salmon_1": u"209",
    u"light_coral": u"210",
    u"pale_violet_red_1": u"211",
    u"orchid_2": u"212",
    u"orchid_1": u"213",
    u"orange_1": u"214",
    u"sandy_brown": u"215",
    u"light_salmon_1": u"216",
    u"light_pink_1": u"217",
    u"pink_1": u"218",
    u"plum_1": u"219",
    u"gold_1": u"220",
    u"light_goldenrod_2b": u"221",
    u"light_goldenrod_2c": u"222",
    u"navajo_white_1": u"223",
    u"misty_rose1": u"224",
    u"thistle_1": u"225",
    u"yellow_1": u"226",
    u"light_goldenrod_1": u"227",
    u"khaki_1": u"228",
    u"wheat_1": u"229",
    u"cornsilk_1": u"230",
    u"grey_100": u"231",
    u"grey_3": u"232",
    u"grey_7": u"233",
    u"grey_11": u"234",
    u"grey_15": u"235",
    u"grey_19": u"236",
    u"grey_23": u"237",
    u"grey_27": u"238",
    u"grey_30": u"239",
    u"grey_35": u"240",
    u"grey_39": u"241",
    u"grey_42": u"242",
    u"grey_46": u"243",
    u"grey_50": u"244",
    u"grey_54": u"245",
    u"grey_58": u"246",
    u"grey_62": u"247",
    u"grey_66": u"248",
    u"grey_70": u"249",
    u"grey_74": u"250",
    u"grey_78": u"251",
    u"grey_82": u"252",
    u"grey_85": u"253",
    u"grey_89": u"254",
    u"grey_93": u"255",
}


def FG_COLOURS(name):
    return u"38;5;{}".format(COLOURS[name])


def BG_COLOURS(name):
    return u"48;5;{}".format(COLOURS[name])
